### ISSUE TITLE
USB MIDI host possible?

### ISSUE THREAD

**Issue by probonopd** [shorepine/amy_dual_core_esp32#1]
Is it possible to connect a USB MIDI controller (keyboard) to the USB plug and play the synthesizer live that way? 

References:
* https://github.com/touchgadget/esp32-usb-host-demos?tab=readme-ov-file#usbhmidi----usb-host-midi
* https://github.com/esp32beans/ESP32USBMIDI

**Comment by sauloverissimo** [comment]
@probonopd Yes, USB MIDI Host is definitely possible on the ESP32-S3. The S3 has a native USB OTG peripheral that supports host mode, so you can plug a USB MIDI keyboard directly via an OTG cable — no extra hardware needed.

For the implementation side, [ESP32_Host_MIDI](https://github.com/sauloverissimo/ESP32_Host_MIDI) is a library that handles USB Host MIDI (among other transports) on the S3. It takes care of enumeration, endpoint handling, and hot-plug. Could be a good starting point if you want to integrate USB MIDI input into AMY without writing the USB stack from scratch.

**Comment by dpwe** [comment]
Note: This repo is mostly subsumed, I believe, by the main AMY repo, which includes Arduino examples for running on many different MCUs, including ESP with dual core.  AMY now includes MIDI handlers, including USB MIDI, for various hardware configurations.  I'm not 100% sure if ESP32 is there or not; if it isn't, that would be the place to add it.

The TulipCC repo includes a micropython wrapper around AMY to run on ESP32 hardware.  It's aimed at our custom TulipCC hardware (includes a touch screen), or the forthcoming AMYboard (ESP32 with on-board ADC, DAC, SPDIF, MIDI).  That code should be pretty easy to adapt to other hardware.

**Comment by sauloverissimo** [comment]
@dpwe Thanks for the context — good to know the main AMY repo is the right place for ESP32 MIDI additions.

For what it's worth, I have USB Host MIDI working on the ESP32-S3 right now with several live examples. Here's a quick overview:

### Piano Visualizer + Music Theory + Synth
Real-time piano UI with chord/interval detection ([Gingoduino](https://github.com/sauloverissimo/gingoduino)) and onboard polyphonic synthesis via I2S DAC. Arturia Minilab 25 plugged directly via USB OTG — no shields, no extra hardware.

![Piano](https://raw.githubusercontent.com/sauloverissimo/ESP32_Host_MIDI/main/examples/T-Display-S3-Piano/images/pianno-MIDI-25.jpeg)

### Piano Debug — MIDI Event Monitor
Raw MIDI byte inspector with event log, dual tracking (queue vs internal state), and mini-piano bar. Useful for debugging USB MIDI issues without a Serial port.

![Piano Debug](https://raw.githubusercontent.com/sauloverissimo/ESP32_Host_MIDI/main/examples/T-Display-S3-Piano-Debug/images/pianno-debug.jpeg)

### BLE MIDI — Wireless Transport
BLE MIDI receiver running simultaneously with USB Host on the same ESP32-S3. Messages from both transports share one event queue.

![BLE MIDI](https://raw.githubusercontent.com/sauloverissimo/ESP32_Host_MIDI/main/examples/T-Display-S3-BLE-Receiver/images/BLE.jpeg)

### RTP-MIDI / Apple MIDI over WiFi
Network MIDI via RTP/UDP — compatible with macOS/iOS MIDI Network and rtpMIDI on Windows.

![RTP-MIDI](https://raw.githubusercontent.com/sauloverissimo/ESP32_Host_MIDI/main/examples/RTP-MIDI-WiFi/images/RTP.jpeg)

### MIDI 2.0 / UMP over UDP
MIDI 2.0 Universal MIDI Packet transport over UDP — one of the first ESP32 implementations.

![MIDI 2.0](https://raw.githubusercontent.com/sauloverissimo/ESP32_Host_MIDI/main/examples/T-Display-S3-MIDI2-UDP/images/MIDI2.jpeg)

### Gingoduino Integration — Music Theory Engine
Real-time music theory analysis on the MCU: chords, scales, intervals, formulas — all computed locally with zero latency.

![Gingoduino](https://raw.githubusercontent.com/sauloverissimo/ESP32_Host_MIDI/main/examples/T-Display-S3-Gingoduino/images/gingo-duino-integration.jpeg)

### Queue Monitor
Live MIDI event queue visualization showing all transports feeding into a single shared queue.

![Queue](https://raw.githubusercontent.com/sauloverissimo/ESP32_Host_MIDI/main/examples/T-Display-S3-Queue/images/queue.jpeg)

---

All of this runs on a LilyGO T-Display-S3. The library ([ESP32_Host_MIDI](https://github.com/sauloverissimo/ESP32_Host_MIDI)) supports 9 transports (USB Host, USB Device, BLE, Apple MIDI/RTP, OSC, ESP-NOW, UART/DIN-5, Ethernet, MIDI 2.0/UMP) through a single API. Works with Arduino IDE, PlatformIO, and ESP-IDF.

The core loop is minimal:

```cpp
void loop() {
    midiHandler.task();
    midiHandler.fillActiveNotes(activeNotes);
    // ... render piano, analyze chords, play synth
}
```

If there's interest in integrating USB Host MIDI into AMY's ESP32 path, I'd be happy to help — the USB Host stack is the tricky part and it's already solved.
